### ISSUE TITLE
[Zenith] `dagger script`: invoke any function from any module, in your favorite language, from the CLI

### ISSUE THREAD

**Issue by shykes** [dagger/dagger#5774]
## Overview

In #5757, we take the Zenith in a bold new direction:
* everything is a function
* functions are packaged together in modules
* each module has an API
* A module's API can be queried either 1) from other functions, or 2) **directly by the end user**

But for this design to work, we need **a great UI for querying a module's API**. More specifically, we need **a great UI for running a module's functions** (using its API).

We think Dagger will support many ways to do this, including:

* Crafting raw graphql queries with `curl`, `postman` or other HTTP/graphql clients
* Crafting a query from CLI arguments and flags, using `dagger do` or an equivalent subcommand of the Dagger CLI
* Using a custom web UI, either running locally, in Dagger Cloud, or in an IDE plugin
* Crafting a query in a familiar programming language, with the help of a specially packaged language interpreter

Of all these options, I propose that we focus primarily on the last (familiar language interpreter), and call it `dagger script`

## `dagger script`: call Dagger functions from the command-line, in a familiar language

This section is work in progress. Bikeshedding is expected and encouraged :)

### Overall architecture

FIXME

### The versioning problem

One unresolved question is how to manage versioning of the various pieces (CLI, engine, SDK), taking into consideration ongoing changes (#5583 , #5484), in a way that 1) makes `dagger script` great to use, 2) doesn't interfere with ongoing design work, 3) ongoing design work doesn't later break `dagger script`.

--> FIXME, add details here. For now, see [this discord thread](https://discord.com/channels/707636530424053791/1148666309324845167/1148710601338654760)

**Comment by helderco** [comment]
This sounds a lot like `dagger run` today, rather than the "one-liner" we discussed a few times, based on the direct access to the user's system. Unless you mean through an API?

**Comment by sipsma** [comment]
Yeah agree reading this sounds more and more like https://github.com/dagger/dagger/issues/5993; if you want to write a script that can call out to your functions, you will need codegen bindings to those functions.

I think if you take that and combine it with the ability for `dagger run` to compile your code for you rather than require host dependencies (i.e. `dagger run main.go` rather than `dagger run go run main.go`) you'd end up with a pretty simple and easy way to write complex logic that executes direct on your host and calls out to functions.

I've been thinking through other variations too in that issue, but something along these lines feels right. I'm hoping to pick up #5993 in the near future, so should be able to start playing around with ideas soon 🤞

**Comment by shykes** [comment]
@helderco @sipsma sorry I didn't have time to finish my writeup. It's different from eg. `dagger run main.go` in a few ways:

- Dagger Scripts have no dependency other than `dagger`.
- Dagger Scripts have access to the user's system, but not *unlimited* access. Access is still mediated by the Dagger Engine, which allows us to implement guardrails, such as a trust system.
- There is one possible implementation of Dagger Scripts which does *not* require codegen: a bash-compatible shell. See https://daggerverse.dev/mod/github.com/shykes/daggerverse/dsh for a POC implementation.

**Comment by shykes** [comment]
@helderco @sipsma update: I added more details to the "Solution" section:

- Comparison of scripts vs. functions vs. external clients
- Fleshed out proposal for one possible design: bash-compatible shell

Next up:

- also flesh out the "python or typescript one-liner" design.
- Discuss if those two designs are complementary, or overlapping (no idea)

**Comment by jedevc** [comment]
A couple of thoughts on this, from the context of modularizing our own CI.

There are a couple of *different* use cases to be aware of, things that I want to do as part of my workflow with our new `ci` directory, so we can get rid of mage:
- Simple aliases. e.g. `dagger script docs-generate` or similar translates to `dagger call --source=. docs generate export --path=.`
- Chaining commands - I want to be able to "pipe" data between different calls - if I get a `Container` from one script, I want to be able to use it as the input for another call. e.g. `dagger script scan` or similar translates to an `dagger call --source=. engine container` being input to `dagger call -m trivy-scanner --input=<container-from-previous-step>`. (this would be a huge improvement to composing modules together).
- Running commands in parallel. e.g. `dagger script generate-all` should run all of `dagger call --source=. sdk <go|python|...> generate export --path=.` all at once, in parallel (today this is a pain, all the output really just overlaps badly).

---

I like the idea of scripts being just SDKs run outside of the engine - this means it's *wayyy* easier to express all of these above things - trying to have something shell-like that can do all this feels a bit annoying to have to write/maintain/use.

But... if we do go that route, potentially we could have a bash-SDK? Or something like that, maybe it could only be used for implementing scripts - that way we could still allow having (very limited) scripts written in bash, paired with the power of dagger. :laughing: 

---

It seems like the question at the moment is whether these scripts execute on the client or on the engine? If they run on the client, it feels like it's much too tempting to manually modify and edit files, instead of using the dagger api to access them - which then is trickier to visualize in the tui/dagger cloud, get good caching, etc - it's also very dependent on the client environment. I'd much rather have the environment be the SDK function containers, even if they're not using exactly the same codegen APIs, etc.

**Comment by shykes** [comment]
> today this is a pain, all the output really just overlaps badly

Wouldn't this be fixed by all the calls sharing the same session? Similar to `dagger run sh`?

**Comment by pjmagee** [comment]
Would introducing bash not seen in some ways leaning towards "glue". I know it isn't with how Dagger works but an argument about not having to have bash scripts for nice clean portable programmable CICD Pipelines seems to have less meaning when a Bash SDK exists. Maybe I'm not fully understanding this and this is likely the case but as an outsider just being curious reading proposals.
